<h1><b><i>Principal Component Analysis</i> (<i>PCA</i>) </b></h1>

<p>Στην άσκηση αυτή θα μελετήσετε τον αλγόριθμο <b><i>ανάλυσης σε κύριες συνιστώσες</i></b> (<b><i>Principal Component Analysis</i></b>, <b><i>PCA</i></b>), υλοποιημένο σύμφωνα με τη <b><i>μέθοδο συνδιακύμανσης</i></b> (<b><i>covariance method</i></b>). Για να κατανοήσετε τη χρησιμότητα της μεθόδου θα εκπαιδεύσετε και θα αξιολογήσετε την ακρίβεια ενός μοντέλου <b><i>logistic regression</i></b> για ένα dataset πριν και μετά την εφαρμογή του αλγορίθμου <b><i>PCA</i></b>. Περισσότερες πληροφορίες για τη μέθοδο αυτή μπορείτε να αναζητήσετε <a href="https://ourarchive.otago.ac.nz/handle/10523/7534">εδώ</a>.</p>

<p>Η άσκηση περιλαμβάνει <b><i>δύο</i></b> προγράμματα <i>Python</i>: (a) το πρώτο δέχεται ένα dataset σε μορφή .<i>csv</i>, εφαρμόζει τον αλγόριθμο <b><i>PCA</i></b> και δημιουργεί το αρχείο <b><i>foo.csv</i></b> με το μετασχηματισμένο dataset, όπως προκύπτει από τις κύριες συνιστώσες που επέλεξε ο χρήστης να διατηρήσει, (b) το δεύτερο δέχεται ένα αρχείο σε μορφή .<i>csv</i>, διαχωρίζει το dataset σε <i>training</i> και <i>test set</i>, εκπαιδεύει ένα μοντέλο <b><i>logistic regression</i></b>, χρησιμοποιώντας το <i>training set</i> και υπολογίζει τον αριθμό των σφαλμάτων του μοντέλου πάνω στο <i>test set</i>.</p>

<p>Το dataset που θα χρησιμοποιήσετε παρέχεται σε δύο μορφές: (a) <b><i><a href="https://raw.githubusercontent.com/nkostopoulos/StochasticsLabPublic/master/lab2/demo3a.csv">demo3a.csv</a></i></b> και (b) <b><i><a href="https://github.com/nkostopoulos/StochasticsLabPublic/blob/master/lab2/demo3b.csv">demo3b.csv</a></i></b>, το οποίο δεν περιλαμβάνει την πρώτη στήλη του <b><i><a href="https://raw.githubusercontent.com/nkostopoulos/StochasticsLabPublic/master/lab2/demo3a.csv">demo3a.csv</a></i></b>, δηλαδή τα <i>labels</i> που αντιστοιχούν σε κάθε είσοδο. Τα datasets αυτά αποτελούν απλοποιημένη μορφή του dataset που μπορεί να βρεθεί <a href="https://archive.ics.uci.edu/ml/datasets/wine">εδώ</a>.</p>

<h3><b><i>Ανάλυση σε Κύριες Συνιστώσες</i></b></h3>
<p>Αρχικά, θα φορτώσετε τις βιβλιοθήκες που απαιτούνται για το πρόγραμμα που θα αναλύσει το dataset <b><i><a href="https://github.com/nkostopoulos/StochasticsLabPublic/blob/master/lab2/demo3b.csv">demo3b.csv</a></i></b> στις κύριες συνιστώσες του.</p>




In [1]:
!pip install numpy
#https://machinelearningmastery.com/calculate-principal-component-analysis-scratch-python/
from numpy import genfromtxt
from numpy import mean
from numpy import cov
from numpy.linalg import eig
import numpy as np

<p>Στη συνέχεια, θα φορτώσετε το dataset <b><i></i></b></p>

In [2]:
data = genfromtxt('https://raw.githubusercontent.com/nkostopoulos/StochasticsLabPublic/master/lab2/demo3b.csv', delimiter=',')

<p>Έπειτα, θα υπολογίσετε το μέσο όρο κάθε στήλης (feature) του dataset και θα κανονικοποιήσετε κάθε feature με αυτόν</p>

In [3]:
M = mean(data.T, axis=1)
data_normal = data - M

<p>Στη συνέχεια, θα υπολογίσετε το <b><i>πίνακα συνδιακύμανσης</i></b> (<b><i>covariance matrix</i></b>) για το dataset</p>

In [4]:
covariance = cov(data_normal.T)
print("The covariance matrix of the normalized data is the following: ")
print(covariance)

The covariance matrix of the normalized data is the following: 
[[ 7.89911157e-01  1.64350328e-02  5.03965534e-02 -1.39557561e+00
   4.51870543e+00  2.35123047e-01  3.56524806e-01 -2.92133512e-02
   9.15320632e-02  1.07205560e+00  3.92238044e-03  1.22678110e-01
   2.36909466e+02]
 [ 1.64350328e-02  7.74344335e-01  3.18039952e-02  4.24165474e-01
  -2.21434109e-01  1.44160704e-02  3.75581395e-02  5.34385808e-03
   6.47891175e-02 -1.54797734e-01 -6.06599761e-02  6.96007454e-02
  -3.33629219e+01]
 [ 5.03965534e-02  3.18039952e-02  8.83052117e-02  3.73006798e-01
   1.40136434e+00  4.12452594e-02  8.26573643e-02  5.81539654e-03
   1.24709839e-02  1.22397007e-01  2.06346094e-03  3.02871079e-02
   3.18556160e+01]
 [-1.39557561e+00  4.24165474e-01  3.73006798e-01  1.15631181e+01
  -6.46193798e+00 -4.24464222e-01 -4.52046512e-01  1.24282469e-01
  -1.77910614e-01 -2.39565069e+00 -2.05387955e-02 -1.02069171e-03
  -5.11872284e+02]
 [ 4.51870543e+00 -2.21434109e-01  1.40136434e+00 -6.46193798e+00
  

<p>Το επόμενο βήμα είναι να υπολογίσετε τις <b><i>ιδιοτιμές</i></b> (<b><i>eigenvalues</i></b>) και τα <b><i>ιδιοδιανύσματα</i></b> (<b><i>eigenvectors</i></b>) του dataset.</p>

In [5]:
values, vectors = eig(covariance)
print("The eigenvalues of the normalized data are the following: ")
print(values)

The eigenvalues of the normalized data are the following: 
[1.24313073e+05 1.93456026e+02 9.50168321e+00 1.32500169e+00
 8.10595669e-01 4.51081713e-01 2.69157831e-01 1.60993380e-01
 1.11100343e-01 6.13399000e-03 2.19782528e-02 3.20312852e-02
 5.87421686e-02]


<p>Στη συνέχεια, θα επιλέξετε τις πιο σημαντικές <b><i>ιδιοτιμές</i></b> και θα προσαρμόσετε αντίστοιχα τα <b><i>ιδιοδιανύσματα</i></b> του dataset.</p>

In [6]:
new_values = values[0:3]
print("The most important eigenvalues are the following: ")
print(new_values)
new_vectors = vectors[0:3]
print("The most important eigenvectors are the following: ")
print(new_vectors)

The most important eigenvalues are the following: 
[1.24313073e+05 1.93456026e+02 9.50168321e+00]
The most important eigenvectors are the following: 
[[-1.90615566e-03 -4.75319302e-04 -4.70065618e-02  2.25255930e-01
   9.82089183e-02 -2.20758056e-01 -8.52483156e-01 -3.65140031e-01
  -1.65925760e-01 -5.62139324e-03 -2.07811823e-02  2.48378617e-02
  -8.18675258e-04]
 [ 2.68379014e-04 -2.11118304e-03  3.21935310e-02  1.25115825e-01
   9.16405147e-01 -3.07909300e-01  1.86306127e-01  7.26209733e-02
  -1.81047372e-02 -1.16925805e-02  8.44468784e-02 -2.41209813e-02
   2.31770029e-02]
 [-2.56407459e-04 -4.21644285e-03  5.19948847e-02  4.67286333e-02
   2.01302236e-02 -3.45064791e-03 -1.57617050e-02  1.52544206e-01
  -7.18891029e-02 -1.38330530e-01 -6.68720003e-02  6.75337422e-01
  -6.97354965e-01]]


<p>Τώρα, θα εφαρμόσετε τα νέα <b><i>ιδιοδιανύσματα</i></b> στο παλιό dataset για να πάρετε το νέο, μειωμένο σε μέγεθος dataset.</p>

In [7]:
new_data = new_vectors.dot(data_normal.T)

<p>Να αποθηκεύσετε το νέο dataset σε ένα αρχείο <i>csv</i>.</p>

In [8]:
np.savetxt("foo.csv", new_data.T, delimiter=",")

<h4><b><i>Ερωτήσεις</i></b></h4>
<ul>
<li>Να μελετήσετε το παραπάνω πρόγραμμα και να περιγράψετε, σύντομα, τα βήματα που ακολουθεί ο αλγόριθμος <b><i>PCA</i></b>, υλοποιημένος με τη μέθοδο <b><i>covariance</i></b>. Να συμπεριλάβετε και τις μαθηματικές πράξεις.</li>
<li>Να εφαρμόσετε τον αλγόριθμο <b><i>PCA</i></b> πάνω στα δεδομένα του αρχείου <i><a href="https://github.com/nkostopoulos/StochasticsLabPublic/blob/master/lab2/demo3b.csv">demo3b.csv</a></i>. Στη συνέχεια, να καταγράψετε τον <i>πίνακα συνδιακύμανσης</i> του dataset και τις ιδιοτιμές του πίνακα αυτού. Τι υποδηλώνουν οι θετικές και τι οι αρνητικές τιμές του <i>πίνακα συνδιακύμανσης</i>; Να διατάξετε τις <i>ιδιοτιμές</i> σε φθίνουσα σειρά. Τι παρατηρείτε για τις τρεις πρώτες σε σχέση με τις υπόλοιπες; Πόσες <i>κύριες συνιστώσες</i> επιλέγει να διατηρήσει ο αλγόριθμος;</li>
</ul>

ΑΠΑΝΤΗΣΕΙΣ

i)Στη συγκεκριμένη άσκηση έχουμε N=130 δειγματικά σημεία, καθ' ένα εκ των οποίων αποτελείται από m=13 χαρακτηριστικά. Έτσι, τα σημεία αυτά καταχωρούνται σε έναν πίνακα X', με διαστάσεις Nxm. Κατόπιν, η μέση τιμή του  i-οστού χαρακτηριστικού,μi, υπολογίζεται ως:

μi=1/N*ΣX'ni, όπου ΣX'i το άθροισμα των σημείων Χ΄ για i από 1 μέχρι n.

με σκοπό να αφαιρεθεί από κάθε δειγματικό σημείο, προκειμένου η μέση τιμή ως προς κάθε χαρακτηριστικό να ισούται με μηδέν: Χ=Χ'-μ

όπου με μ συμβολίζεται ο Nxm πίνακας που σε κάθε γραμμή του έχει το διάνυσμα μ=(μ1,μ2,...μn) . Χρησιμοποιώντας το νέο πίνακα X υπολογίζεται ο αντίστοιχος πίνακας συνδιακύμανσης των χαρακτηριστικών, δηλαδή ο mxm πίνακας , του οποίου το στοιχείο Cij αντιστοιχεί στη συνδιακύμανση μεταξύ του  i-οστού και  j-οστού χαρακτηριστικού:

cij=1/(N-1)Σ(Xni-μi)(Xnj-μj), όπου Σ(Xni-μi)(Xnj-μj) το άθροισμα (Xni-μi)(Xnj-μj) για i από 1 μέχρι n.  

Προφανώς, στην περίπτωση i=j προκύπτει η διακύμανση του i-οστού χαρακτηριστικού. Με δεδομένο τον πίνακα συνδιακύμανσης, το επόμενο βήμα είναι ο υπολογισμός των ιδιοδιανυσμάτων του, x, και των ιδιοτιμών του,λ, λύνοντας το πρόβλημα:

Cx=λx

Ο λόγος για τον οποίο είναι απαραίτητος ο προσδιορισμός των ιδιοδιανυσμάτων και των ιδιοτιμών του πίνακα συνδιακύμανσης είναι πως, για ένα πρόβλημα m χαρακτηριστικών, το σφάλμα προσέγγισης PCA μέσω των
k κύριων συνιστωσών, J, θα ισούται με:

J=Σλi, όπου Σλi το άθροισμα λi για i από k+1 μέχρι m.






ii)Ο πίνακας συνδιακύμανσης είναι ένας σημαντικός υπολογιστικός πίνακας που χρησιμοποιείται στη στατιστική και στην ανάλυση δεδομένων. Οι θετικές και αρνητικές τιμές του πίνακα συνδιακύμανσης υποδηλώνουν τον τύπο της σχέσης μεταξύ δύο μεταβλητών. Ειδικότερα,η συνδιακύμανση είναι ένα μέτρο του πόσο σχετίζονται οι παραλλαγές δύο μεταβλητών. Μια θετική συνδιακύμανση μεταξύ δύο μεταβλητών αποκαλύπτει ότι οι ζευγαρωμένες τιμές και των δύο μεταβλητών τείνουν να αυξάνονται μαζί. Μια αρνητική συνδιακύμανση αποκαλύπτει ότι υπάρχει αντίστροφη σχέση μεταξύ των μεταβλητών, δηλαδή, καθώς η μία αυξάνεται, η άλλη τείνει να μειώνεται. Μια μηδενική συνδιακύμανση δείχνει ότι δεν υπάρχει σύνδεση μεταξύ των τιμών των δύο μεταβλητών

In [9]:
print("Οι ιδιοτιμές του πίνακα συνδιακύμανσης σε φθίνουσα σειρά είναι οι εξής:")
print(sorted(values,reverse=True))

Οι ιδιοτιμές του πίνακα συνδιακύμανσης σε φθίνουσα σειρά είναι οι εξής:
[124313.07344459958, 193.45602626538692, 9.50168321005974, 1.3250016880448792, 0.8105956686469435, 0.4510817126522554, 0.26915783148547284, 0.16099338038871566, 0.11110034340026094, 0.0587421685521958, 0.03203128519816791, 0.021978252812754256, 0.006133989998101028]


Οι 3 πρώτες τιμές έχουν πολύ μεγάλη απόσταη από τις υπόλοιπες. Συγκεκριμένα, οι 3 πρώτες τιμές έχουν τιμή: 124313.07344459958,
193.45602626538692, 9.50168321005974. Από εκεί και πέρα οι τιμές που υπάρχουν είναι  1.3250016880448792 και οι υπόλοιπες τιμές είναι μικρότερες του 1. Συνεπώς, ύπαρχει σφάλμα.

 Οι κύριες συνιστώσες που επιλέγει ο αλγόριθμος είναι οι 3 πρώτες.Αυτό γίνεται προκειμένου να μειώσουμε το σφάλμα.Η ιδέα είναι ότι κρατώντας τις ιδιοτιμές που βρίσκονται πιο κοντά μεταξύ τους και έχουν μεγάλη απόσταση από τις υπόλοιπες, το σφάλμα προσέγγισης θα είναι λογικά μικρότερο. Το σφάλμα προσέγγισης της PCA για k κύριες συνιστώσες δίνεται από το άθροισμα των k-d μικρότερων ιδιοτιμών, θεωρώντας πως έχουν κανονικοποιηθεί. Στην προκείμενη περίπτωση, όπου έχουν επιλεγεί 3 κύριες συνιστώσες, το σφάλμα αυτό θα ισούται με:

In [10]:
print(f"Το σφάλμα προσέγγισης 3 κυρίων συνιστωσών ισούται με {100*sum(values[3:])/sum(values):.4f}%")

Το σφάλμα προσέγγισης 3 κυρίων συνιστωσών ισούται με 0.0026%


<h3><b><i>Logistic Regression</i></b></h3>

<p>Αρχικά, θα φορτώσετε τις απαραίτητες βιβλιοθήκες.</p>

In [112]:
!pip install pandas
!pip install numpy
!pip install matplotlib
!pip install sklearn
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
from sklearn.model_selection import train_test_split
from sklearn.decomposition import PCA
from sklearn.preprocessing import StandardScaler
from sklearn.linear_model import LogisticRegression

  Using cached sklearn-0.0.post12.tar.gz (2.6 kB)
  error: subprocess-exited-with-error
  
  × python setup.py egg_info did not run successfully.
  │ exit code: 1
  ╰─> See above for output.
  
  note: This error originates from a subprocess, and is likely not a problem with pip.
  Preparing metadata (setup.py) ... error
error: metadata-generation-failed

× Encountered error while generating package metadata.
╰─> See above for output.

note: This is an issue with the package mentioned above, not pip.
hint: See above for details.


Στη συνέχεια, θα φορτώσετε τα datasets που θα χρησιμοποιηθούν για την εκπαίδευση του μοντέλου <b><i>logistic regression</i></b>. Στην πρώτη περίπτωση, θα εκπαιδεύσετε το μοντέλο, χρησιμοποιώντας το αρχείο <i><a href="https://raw.githubusercontent.com/nkostopoulos/StochasticsLabPublic/master/lab2/demo3a.csv">demo3a.csv</a></i>. Στη δεύτερη περίπτωση, θα φορτώσετε το αρχείο <i>foo.csv</i> που πήρατε ως έξοδο από το προηγούμενο τμήμα κώδικα της άσκησης (<i>Ανάλυση σε Κύριες Συνιστώσες</i>), προσθέτοντας στην πρώτη στήλη του αρχείου τα labels που υπάρχουν στην πρώτη στήλη του αρχείου <a href="https://raw.githubusercontent.com/nkostopoulos/StochasticsLabPublic/master/lab2/demo3a.csv"><i>demo3a.csv</i></a>.

In [81]:
df = pd.read_csv("https://raw.githubusercontent.com/nkostopoulos/StochasticsLabPublic/master/lab2/demo3a.csv")
#df = pd.read_csv("foo.csv")

<p>Έπειτα, θα χωρίσετε το dataset σε <i>training</i> και <i>test set</i>.</p>

In [82]:
# Separate the input features from the target variable
x = df.iloc[:,1:13].values
y = df.iloc[:,0].values
# Split the dataset into train and test set
Xtrain,Xtest,Ytrain,Ytest = train_test_split(x,y, test_size = 0.2)

<p>Έπειτα, θα εκπαιδεύσετε το μοντέλο <b><i>Logistic Regression</i></b></p>

In [83]:
%%time
classifier = LogisticRegression(max_iter = 1000)
classifier.fit(Xtrain,Ytrain)

CPU times: user 76.7 ms, sys: 0 ns, total: 76.7 ms
Wall time: 88.3 ms


LogisticRegression(max_iter=1000)

<p>Τέλος, θα λάβετε τις προβλέψεις του μοντέλου που εκπαιδεύσατε πάνω στο test set και θα υπολογίσετε τα συνολικά λάθη.</p>

In [84]:
# Get the predictions on the test set
prediction = classifier.predict(Xtest)

# Calculate the total number of errors on the test set
errors = 0
for index in range(0,len(prediction) - 1):
	if prediction[index] != Ytest[index]:
		errors += 1

print("Total errors on the test dataset")
print(errors)

Total errors on the test dataset
1


Παρατηρούμε ότι υπάρχει 1 λάθος. Το μοντέλο μας δεν είναι το βέλτιστο, άλλα παρ΄όλα αυτά βρίσκεται σε πολύ καλό επίπεδο εγκυρότητας.

Σε αυτό το σημείο επαναλαμβάνουμε την ίδια διαδικασία, αλλά αυτή τη φορά διαβάζουμε το αρχείο "foo.csv"

In [101]:
df = pd.read_csv("foo.csv")
#print(df)
x=df.values
#print(x)
Xtrain,Xtest,Ytrain,Ytest = train_test_split(x,y, test_size = 0.2)

In [102]:
%%time
classifier = LogisticRegression(max_iter = 1000)
classifier.fit(Xtrain,Ytrain)

CPU times: user 8.01 ms, sys: 973 µs, total: 8.98 ms
Wall time: 9 ms


LogisticRegression(max_iter=1000)

In [103]:
# Get the predictions on the test set
prediction = classifier.predict(Xtest)

# Calculate the total number of errors on the test set
errors = 0
for index in range(0,len(prediction) - 1):
	if prediction[index] != Ytest[index]:
		errors += 1

print("Total errors on the test dataset")
print(errors)

Total errors on the test dataset
1


Τώρα θα ελέγξουμε τι γίνεται στην περίπτωση που κρατάμε 1 και 2 κύριες συνιστώσες.


Αρχικά θα ελέγξουμε τι γίνεται στην περίπτωση όπου κρατάμε 2 συνιστώσες.

In [104]:
df = pd.read_csv("foo.csv")
x = df.iloc[:,1:3].values
Xtrain,Xtest,Ytrain,Ytest = train_test_split(x,y, test_size = 0.2)

In [105]:
%%time
classifier = LogisticRegression(max_iter = 1000)
classifier.fit(Xtrain,Ytrain)

CPU times: user 8.11 ms, sys: 0 ns, total: 8.11 ms
Wall time: 10.7 ms


LogisticRegression(max_iter=1000)

In [106]:
# Get the predictions on the test set
prediction = classifier.predict(Xtest)

# Calculate the total number of errors on the test set
errors = 0
for index in range(0,len(prediction) - 1):
	if prediction[index] != Ytest[index]:
		errors += 1

print("Total errors on the test dataset")
print(errors)

Total errors on the test dataset
0


Τώρα θα ελέγξουμε τη περίπτωση όπου κρατάμε 1 συνιστώσα.

In [107]:
df = pd.read_csv("foo.csv")
x = df.iloc[:,2:3].values
Xtrain,Xtest,Ytrain,Ytest = train_test_split(x,y, test_size = 0.2)

In [108]:
%%time
classifier = LogisticRegression(max_iter = 1000)
classifier.fit(Xtrain,Ytrain)

CPU times: user 6.61 ms, sys: 1 µs, total: 6.61 ms
Wall time: 7.61 ms


LogisticRegression(max_iter=1000)

In [109]:
# Get the predictions on the test set
prediction = classifier.predict(Xtest)

# Calculate the total number of errors on the test set
errors = 0
for index in range(0,len(prediction) - 1):
	if prediction[index] != Ytest[index]:
		errors += 1

print("Total errors on the test dataset")
print(errors)

Total errors on the test dataset
0


<h4><b><i>Ερώτηση</i></b></h4>
<p>Να εκτελέσετε παραπάνω τμήματα κώδικα με είσοδο τα αρχεία (a) <i><a href="https://raw.githubusercontent.com/nkostopoulos/StochasticsLabPublic/master/lab2/demo3a.csv">demo3a.csv</a></i> και (b) <i>foo.csv</i>. Τι παρατηρείτε για την ακρίβεια του μοντέλου στις δύο περιπτώσεις; Δοκιμάστε και για την περίπτωση που κρατάμε (α) 1 και (β) 2 <i>κύριες συνιστώσες</i>. Τι παρατηρείτε;</p>

<h3><b><i>Συμπληρωματικές Ερωτήσεις</i></b></h3>
<ul>
<li>Ποια είναι η χρησιμότητα του αλγορίθμου <b><i>PCA</i></b> ως προς τη δυνατότητα <i>οπτικοποίησης</i> (<i>visualization</i>) των δεδομένων του dataset;</li>
<li>Ποια είναι η χρησιμότητα του αλγορίθμου <b><i>PCA</i></b> ως προς την ταχύτητα εκπαίδευσης του μοντέλου logistic regression; Να βασίσετε την απάντησή σας στα δύο παρακάτω τμήματα κώδικα.</li>
</ul>

ΑΠΑΝΤΗΣΕΙΣ:


i) Ο αλγόριθμος PCA  (Principal Component Analysis) είναι ένα ισχυρό εργαλείο για τη μείωση των διαστάσεων των δεδομένων και μπορεί να χρησιμοποιηθεί για τη δημιουργία μιας νέας σειράς χαρακτηριστικών (principal components) που αντιπροσωπεύουν τις κύριες κατευθύνσεις της μεγάλης διακύμανσης στα δεδμένα. Η χρησιμότητα του PCA στην οπτικοποίηση των δεδομένων προκύπτει από το γεγονός ότι μετά τη μείωση της διάστασης, είναι πιο εύκολο να αναπαραστήσουμε γραφικά τα δεδομένα σε έναν χαμηλότερης διάστασης χώρο.Ο αλγόριθμος επιτρέπει τη πιο αποτελεσματική οπτικοποίηση των δεδομένων.

ii)Γενικά,η χρησιμότητα του αλγορίθμου PCA ως προς την ταχύτητα εκπαίδευσης του μοντέλου logistic regression σχετίζεται με τη μείωση της διάστασης των δεδομένων.

Κατά την εκπαίδευση ενός μοντέλου logistic regression, το πρόβλημα της αύξησης της διάστασης των δεδομένων (όταν έχουμε πολλά χαρακτηριστικά) μπορεί να οδηγήσει σε μεγάλο χρόνο εκπαίδευσης και σε προβλήματα υπερ-εκπαίδευσης (overfitting). Ο αλγόριθμος PCA μειώνει τη διάσταση των δεδομένων κρατώντας τη σημαντική πληροφορία και αφαιρώντας την περιττή, καθιστώντας έτσι την εκπαίδευση πιο γρήγορη.

Συγκεκριμένα, οι βασικοί τρόποι μείωσης της διάστασης μέσω του PCA είναι η εύρεση των κύριων συνιστωσών (principal components) και η διατήρηση μόνο των κυριότερων αυτών συνιστωσών. Αυτό οδηγεί σε μια πιο αποδοτική αναπαράσταση των δεδομένων, καθώς το μοντέλο logistic regression εκπαιδεύεται σε ένα χώρο χαμηλότερης διάστασης, χωρίς να χάνεται σημαντική πληροφορία.

Αυτό επιβεβαίωνεται και από την εκτέλεση του παραπάνω κώδικα στο αρχείο "foo.csv". Όταν εκτελούμε τον αλγόριθμο διατηρώντα όλες τις συνιστώσες, τα λάθη είναι 2. Όταν κρατάμε 2 συνιστώσες  1 συνιστώσα , δεν υπάρχει  λάθος. Χρησιμοποιώντας όλες τις συνιστώσες χρειάστηκε 8.1 ms, ενώ για μία κύρια συνιστώσα χρειάστηκε 6.61 ms.